portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -36,37 +36,45 @@
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    Query query = new Query("Comment").addSort("rate", SortDirection.DESCENDING);
-
-    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
-    PreparedQuery results = datastore.prepare(query);
-
-    List<Comment> comments = new ArrayList<>();
-    // Add all the comments in the datastore to the comments list
-    for (Entity entity : results.asIterable()) {
-        String authorVal = (String) entity.getProperty("author");
-        int rateVal = (int)(long) entity.getProperty("rate");
-        ArrayList<String> likedOptionsVal = new ArrayList<>();
-        if (Boolean.parseBoolean((String) entity.getProperty("is_info_liked"))){
-           likedOptionsVal.add("The info");
-        }
-        if (Boolean.parseBoolean((String) entity.getProperty("is_facts_liked"))){
-            likedOptionsVal.add("The facts");
-        }
-        if (Boolean.parseBoolean((String) entity.getProperty("is_gallery_liked"))){
-            likedOptionsVal.add("The gallery");
-        }
-        if (Boolean.parseBoolean((String) entity.getProperty("is_other_liked"))){
-            likedOptionsVal.add("Other");
+      Query query = new Query("Comment").addSort("timestamp", SortDirection.DESCENDING);
+      
+      DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+      PreparedQuery results = datastore.prepare(query);
+      
+      // Number of comments to display 
+      int limit = Integer.parseInt(request.getParameter("limit"));
+      List<Comment> comments = new ArrayList<>();
+      // Add the number of comments specified by the user to the comments list
+      for (Entity entity : results.asIterable()) {
+          if (comments.size()<limit) {
```

Comment:
This is kind of wasteful, you're fetching all the rows.
The limiting should be at the lowest level possible (when the query is being executed).

Hint: use the PreparedQuery#asIterable method with the FetchOptions argument.